File: C++/self_dividing_numbers.cpp
#include <iostream>
#include "./self_dividing_numbers.h"
#include "./utility.h"

using namespace std;

vector<int> Solution::selfDividingNumbers(int left, int right) {
    vector<int> output;
    int integer = left;
    bool is_self_dividing;
    while (integer <= right) {
        is_self_dividing = true;

        // find every digit from the rightmost number
        int base = 10;
        int temp = integer;
        do {
            int rightmost = temp % base;
            if (rightmost == 0 || integer % rightmost != 0) {
                is_self_dividing = false;
                break;
            }
            // shift number
            temp /= base;
        } while(temp != 0);
        
        if (is_self_dividing) output.push_back(integer);
        integer++;
    }
    return output;
}

int main() {
    Solution* sol = new Solution();
    vector<int> output = sol->selfDividingNumbers(1, 22);
    delete sol;
    Utility::printVector(output);
    return 0;
}

File: Python2/self_dividing_numbers.py
class Solution(object):
    def selfDividingNumbers(self, left, right):
        output = []
        integer = left
        base = 10
        while integer <= right:
            is_self_dividing = True
            temp = integer
            while temp:
                rightmost = temp % base
                if rightmost == 0 or integer % rightmost != 0:
                    is_self_dividing = False
                    break
                temp /= base
            if is_self_dividing:
                output.append(integer)
            integer += 1
        return output


if __name__ == '__main__':
    sol = Solution()
    print sol.selfDividingNumbers(1, 22)
